src/renderer/decoder.js

```diff
@@ -115,75 +115,144 @@ function decodePolygon (geometry) {
+                nextPoint = [lineString[i], lineString[i + 1]];
+                nextNormal = getLineNormal(currentPoint, nextPoint);
+                // `turnLeft` indicates that the nextLine turns to the left
+                // `joinNormal` contains the direction and size for the `miter` vertex
+                //  If this is not defined means that the join must be `bevel`.
+                let {turnLeft, joinNormal} = getJoinNormal(prevNormal, nextNormal);
 
-                normals.push(na[0], na[1]);
-                normals.push(nb[0], nb[1]);
-                normals.push(-nb[0], -nb[1]);
+                // Third triangle
+                addTriangle(
+                    [currentPoint, currentPoint, currentPoint],
+                    [[0, 0],
+                        turnLeft ? prevNormal : neg(nextNormal),
+                        turnLeft ? nextNormal : neg(prevNormal)]
+                );
 
-                vertices.push(a[0], a[1]);
-                vertices.push(b[0], b[1]);
-                vertices.push(b[0], b[1]);
+                if (joinNormal) {
+                    // Forth triangle
+                    addTriangle(
+                        [currentPoint, currentPoint, currentPoint],
+                        [joinNormal,
+                            turnLeft ? nextNormal : neg(prevNormal),
+                            turnLeft ? prevNormal : neg(nextNormal)]
+                    );
+                }
             }
-        });
-        breakpoints.push(vertices.length);
-    });
+
+            // Update the variables for the next iteration
+            prevPoint = currentPoint;
+            currentPoint = nextPoint;
+            prevNormal = nextNormal;
+        }
+    }
+
+    function addTriangle (p, n) {
+        vertices.push(p[0][0], p[0][1]);
+        vertices.push(p[1][0], p[1][1]);
+        vertices.push(p[2][0], p[2][1]);
+        normals.push(n[0][0], n[0][1]);
+        normals.push(n[1][0], n[1][1]);
+        normals.push(n[2][0], n[2][1]);
+    }
+}
+
+/**
+ * Compute the normal of a line AB.
+ * By definition it is the unitary vector from B to A, rotated +90 degrees
+ */
+function getLineNormal (a, b) {
+    const u = normalize(vector(b, a));
+    return [-u[1], u[0]];
+}
+
+/**
+ * Compute the normal of the join from the lines' normals.
+ * By definition this is the sum of the unitary vectors `u` (from B to A) and `v` (from B to C)
+ * multiplied by a factor of `1/sin(theta)` to reach the intersecction of the wide lines.
+ * Theta is the angle between the vectors `v` and `u`. But instead of computing the angle,
+ * the `sin(theta)` (with sign) is obtained directly from the vectorial product of `v` and `u`
+ */
+function getJoinNormal (prevNormal, nextNormal) {
+    const u = [prevNormal[1], -prevNormal[0]];
+    const v = [-nextNormal[1], nextNormal[0]];
+    const sin = v[0] * u[1] - v[1] * u[0];
+    const factor = Math.abs(sin);
+    const miterJoin = factor > 0.866; // 60 deg
```

Comment:
With `sin` only we don't have enough information for choosing miter or bevel: factor is small when the angle is close to zero, but also when it's close to 180º.
For the latter I guess the right choice is using miter (although the difference between bevel and mitter for those small changes of direction might be undetectable, were you accounting for that?)

Reply:
Miter is the default, and bevel only used for small angles.

Reply:
But also angles like these:

![screen shot 2018-07-30 at 14 57 34](https://user-images.githubusercontent.com/5909/43398640-faa2f7a2-9408-11e8-8210-2285da3a00ad.png)

In my branch I ended up computing also the cosine to detect these cases: https://github.com/CartoDB/carto-vl/pull/765/files#diff-295612d0417d7be833afc41c3e8a2f63R294

Reply:
After talking with @davidmanzanares this is probably not worth taking care of: the case depicted is is where the lack of miter is most noticeable; for angles closer to 180º it's hardly visible, so it might not be worth adding the miter triangle in these cases (angles between 60 and 180).

Reply:
There is a fallback to bevel joins for <60º. And if we add bevel also for >60º we cover all the angles with bevel. I think we can fallback bevel also for >160º, for example.

Reply:
Hi @jgoizueta. I have reviewed your proposal `!(factor < 0.866 && cos > 0.5);` and saw that this covers better the angle detection for near 180º, so I'm going to use it in this branch. Thanks.